xtask/src/main.rs

```diff
@@ -0,0 +1,148 @@
+
+fn main() {
+    if let Err(e) = try_main() {
+        eprintln!("{}", e);
+        std::process::exit(-1);
+    }
+}
+
+fn try_main() -> Result<(), DynError> {
+    let task = env::args().nth(1);
+    match task.as_deref() {
+        Some("build") => build()?,
+        _ => print_help(),
+    }
+    Ok(())
+}
+
+fn print_help() {
+    eprintln!(
+        "Tasks:
+
+wasm            build wasm using wasm32-wasi target
+components      build components using wasm-tools
+"
+    )
+}
+
+fn build() -> Result<(), DynError> {
+    let _ = fs::remove_dir_all(&target_dir());
+    fs::create_dir_all(&target_dir())?;
+
+    build_wasm()?;
+    build_components()?;
+
+    Ok(())
+}
+
+fn build_wasm() -> Result<(), DynError> {
+    for guest_name in guest_names() {
+        let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+        let status = Command::new(cargo)
+            .current_dir(project_root())
+            .args(&[
+                "build",
+                "--target=wasm32-wasi",
+                "--release",
+                &format!("--package={}", guest_name),
+            ])
+            .status()?;
+
+        if !status.success() {
+            Err(format!(
+                "`cargo build --target wasm32-wasi --release --package {}` failed. Status code: {}",
+                guest_name,
+                status.code().unwrap()
+            ))?;
+        }
+    }
+
+    Ok(())
+}
+
+fn build_components() -> Result<(), DynError> {
+    fs::create_dir_all(project_root().join("components"))?;
+
+    for guest_name in guest_names() {
+        build_component(guest_name)?;
+    }
+
+    Ok(())
+}
+
+fn build_component(name: String) -> Result<(), DynError> {
+    let wasm_tools = env::var("WASM_TOOLS").unwrap_or_else(|_| "wasm-tools".to_string());
+
+    let path_wasm_guest = format!(
+        "{}/wasm32-wasi/release/{}.wasm",
+        target_dir().display(),
+        name
+    );
+
+    let path_wasi_snapshot = format!(
+        "--adapt={}/wasi_snapshot_preview1.reactor.wasm.dev",
+        project_root().display()
+    );
+
+    let path_component_output = format!(
+        "-o={}/components/{}-component.wasm",
+        project_root().display(),
+        name
+    );
+
+    let status = Command::new(wasm_tools)
+        .current_dir(project_root())
+        .args(&[
+            "component",
+            "new",
+            &path_wasm_guest,
+            &path_wasi_snapshot,
+            &path_component_output,
+        ])
+        .status()?;
+
+    if !status.success() {
+        Err(format!(
+            "`wasm-tools component new {} --adapt {} -o {}` failed. Status code: {}",
+            path_wasm_guest,
+            path_wasi_snapshot,
+            path_component_output,
+            status.code().unwrap()
+        ))?;
+    }
+
+    Ok(())
+}
+
+fn guest_names() -> Vec<String> {
+    let path_guests = project_root().join("guests/");
+    let paths = fs::read_dir(path_guests.clone()).unwrap();
+    return paths
```

Comment:
Two things! 

`read_dir` returns all entries in the folder. So, if we add a `guests/README.md` this will pick it up. The `DirEntry` name is a bit misleading.. 

```rust
let paths = ..;

let mut directories = vec![];

for path in paths {
  let path = path.unwrap();
  // use https://doc.rust-lang.org/std/fs/struct.DirEntry.html#method.file_type to get filetype and check `is_dir`
  // push path to directories
}
```

Next, I think we can skip the display, to_string calls. `Path` is a powerful type in rust. Pretty sure we can do something like

```
let status = Command::new(cargo)
            .current_dir(<PATH>)
```

Reply:
Great call! Also played around with is_dir in an early stage, but dropped it as I moved things around

Reply:
Pushed a small refactor. Played around with std::path a bit, but unsure how I could further optimize this code to reduce the occurrences of `.display().to_string()`

We need the strings as output for the `Err`s, so I cannot see how to further reduce them. Could you nudge me in the right direction here?

Reply:
p.s. should I resort to `for of` loops again? There's still a lot of soy in my Rust I reckon? :)